### ISSUE TITLE
Avoiding running the native compiler / vm twice?

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#17870]
Given a value `x`, is there a way to get the result of `Eval native_compute in x` together with a proof that this result is equal to `x`, without needing to run the native compiler twice? 

In the codebase I'm working in, it already takes about 10 minutes to do `Definition y := Eval native_compute in x.`, I don't want to have to spend another 10 minutes in the native compiler again.

If there's not currently a way to do this, I'd like to request a variant of the `vm_compute` and `native_compute` reduction engines that, given a term `x` of type `T`, instead of returning the reduction of `x`, it should compute the reduction `rx` of `x : T` and then return the term `rx <: T` or `rx <<: T`, *without* re-running the reduction of the cast node.

While even in this case it would still be a hassle to avoid running the vm/native compiler twice, it would at least be possible:
```coq
From Coq Require Import NArith.
Set Implicit Arguments.
Fixpoint fib (n : nat) : N := match n with O => 1 | S O => 1 | S (S n'' as n') => fib n' + fib n'' end%N.
Time Eval cbv in fib 30. (* 2.9s *)
Time Eval vm_compute in fib 30. (* 0.125 s *)
Inductive value_eq A : A -> Type := the_value (x : A) : value_eq x.
Definition extract {A x} (val : value_eq x) : A := match val with the_value x => x end.
Lemma extract_eq {A x} (val : @value_eq A x) : extract val = x.
Proof. destruct val; reflexivity. Qed.
Time Definition foo : value_eq (fib 30) := the_value _. (* 0s *)
Time Eval vm_compute in foo. (* 0.135 s *)
Time Definition bar : value_eq (fib 30) := Eval vm_compute in foo. (* 9.166 secs *)
Time Definition bar_val := Eval hnf in extract bar. (* 0s *)
Time Definition bar_eq : bar_val = fib 30 := extract_eq bar. (* 0s *)
```
I'd like something like `Eval vm_compute_with_cast in foo` to make the definition of `bar` be ~0.125s.

**Comment by JasonGross** [comment]
Note, btw, that it is possible to do better than the above.  The following uses the vm for conversion, but crucially still runs the vm twice, which I want to avoid:
```coq
Set Implicit Arguments.
Inductive value_eq A : A -> Type := the_value (x : A) : value_eq x.
Definition extract {A x} (val : value_eq x) : A := match val with the_value x => x end.
Lemma extract_eq {A x} (val : @value_eq A x) : extract val = x.
Proof. destruct val; reflexivity. Qed.
Notation native_compute x := (match x return value_eq x with y => ltac:(native_compute; constructor) end)
                               (only parsing).
Notation vm_compute x := (match x return value_eq x with y => ltac:(vm_compute; constructor) end)
                           (only parsing).


From Coq Require Import NArith.
Fixpoint fib (n : nat) : N := match n with O => 1 | S O => 1 | S (S n'' as n') => fib n' + fib n'' end%N.
Time Eval cbv in fib 30. (* 2.9s *)
Time Eval vm_compute in fib 30. (* 0.125 s *)
Time Definition bar := vm_compute (fib 30). (* 0.27 s *)
Time Definition bar_val := Eval hnf in extract bar. (* 0s *)
Time Definition bar_eq : bar_val = fib 30 := extract_eq bar. (* 0s *)
```

**Comment by SkySkimmer** [comment]
This is wontfix, reduction happens outside the kernel. The kernel only knows conversion.

**Comment by silene** [comment]
Just for the record (presumably longer lasting and more discoverable than [Zulip](https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/Avoiding.20running.20the.20vm.20.2F.20native.20compiler.20twice)), here is a solution Instead of doing
```coq
Definition fib32 := Eval vm_compute in fib 32. (* costly *)
Lemma fib32_spec : fib32 = fib 32.
Proof. vm_cast_no_check (eq_refl fib32). Qed. (* costly again *)
```
one can do
```coq
Definition fib32_def := fib 32.
Lemma fib32_aux : fib32_def = fib 32.
Proof. reflexivity. Qed.
Definition fib32 := Eval vm_compute in fib32_def. (* costly *)
Lemma fib32_spec : fib32 = fib 32.
Proof. rewrite <- fib32_aux. vm_cast_no_check (eq_refl fib32_def). Qed. (* now free *)
```

The idea is that, once the value is given a toplevel name, it will be reduced only once. It would obviously be nicer if Coq was doing this whole dance for the user.
